File: Pod/Classes/Terminals/C2X/HpsC2xDevice.m

#import "HpsC2xDevice.h"
#import "C2XEnums.h"
#import "HpsTerminalResponse.h"
#import <CoreBluetooth/CoreBluetooth.h>

@implementation HpsC2xDevice

- (instancetype)initWithConfig:(HpsConnectionConfig*)config {
    if (self = [super init]) {
        self.config = config;
        self.peripherals = [NSMutableArray new];
    }
    return self;
}

-(void)initialize
{
    if (_centralManager == nil) {
               _centralManager = [[CBCentralManager alloc] initWithDelegate:self queue:nil];
           }

}

- (void)centralManagerDidUpdateState:(CBCentralManager *)central
{
    NSString *currentState = @"";
    switch (central.state)
    {
        case CBManagerStateUnknown:
            currentState = @"CBManagerStateUnknown";
            break;

        case CBManagerStateResetting:
            currentState = @"CBManagerStateResetting";
            break;

        case CBManagerStateUnsupported:
            currentState = @"CBManagerStateUnsupported";
            break;

        case CBManagerStateUnauthorized:
            currentState = @"CBManagerStateUnauthorized";
            break;

        case CBManagerStatePoweredOff:
        {
            currentState = @"CBManagerStatePoweredOff";
            break;
        }

        case CBManagerStatePoweredOn:
            currentState = @"CBManagerStatePoweredOn";            
            [self scan];
            break;
    }

    NSLog(@"State - %@", currentState);
}
- (void)scan
{
    NSDictionary *options = [NSDictionary dictionaryWithObjectsAndKeys:[NSNumber numberWithBool:NO], CBCentralManagerScanOptionAllowDuplicatesKey, nil];
    [self.centralManager scanForPeripheralsWithServices:nil options:options];

    NSLog(@"Scanning started");
}
- (void)centralManager:(CBCentralManager *)central
 didDiscoverPeripheral:(CBPeripheral *)peripheral
     advertisementData:(NSDictionary<NSString *,id> *)advertisementData
                  RSSI:(NSNumber *)RSSI
{
     NSLog(@"%@",peripheral.name);

    if (peripheral.name != nil && ![self.peripherals containsObject:peripheral]) {
        [self.peripherals addObject:peripheral];
    }
    if (self.deviceDelegate!= nil && [self.deviceDelegate respondsToSelector:@selector(onBluetoothDeviceList:)])
    {
        [self.deviceDelegate onBluetoothDeviceList:self.peripherals];
    }
}
- (void)connectDevice:(CBPeripheral *)device
{
    [self.terminalConifguration setHost:device.name];
    
    if ([self isTransactionManagerConnected]) {
        return;
    }
      
    if (self.transactionManager == nil)
    {
        self.transactionManager = [TransactionManager sharedInstance];
    }
    [self.terminalConifguration setConnectionType:BLUETOOTH];
    
    [self initializeTransactionManagerwithSelectedDevice:device];
    
    NSLog(@"%d",self.transactionManager.isInitialized);
    
    if ([self.transactionManager isInitialized]) {
        [self.transactionManager connect:self];
    }
}
 -(void)stopScan
{
    [self.centralManager stopScan];
}
-(BOOL)isTransactionManagerConnected
{
    return (self.transactionManager != nil && [self.transactionManager isConnected]);
}

-(void)disConnect
{
    if ([self isTransactionManagerConnected]) {
        [self.transactionManager disconnect];
    }
}

-(void)getDeviceInfo
{
    if ([self isTransactionManagerConnected]) {
        [self.transactionManager deviceInfo:self];
    }
}
        
-(void)initializeTransactionManagerwithSelectedDevice:(CBPeripheral *)device
{
    self.transactionConfiguration = [self.config transactionConfiguration];
    self.gatewayCofiguration = [self.config gatewayConfig:GatewayTypePortico];
    self.terminalConifguration = [self.config terminalConfig:BBPOS_CHIPPER_2X
                                                                 withTerminalName:device.name];
    [self.transactionManager initializeWithTerminalConfiguration:self.terminalConifguration
                                               transactionConfiguration:self.transactionConfiguration
                                                   gatewayConfiguration:self.gatewayCofiguration];
    
   
}

-(void)processTransactionWithRequest:(TransactionRequest*)transactionRequest
{
    if (self.transactionManager == nil) {
         self.transactionManager = [TransactionManager sharedInstance];
    }
    
    if (![self.transactionManager isInitialized]) {
        //[self initializeTransactionManager];
    }
    
    if ([self.transactionManager isInitialized]) {
          [self.transactionManager startTransaction:transactionRequest withTransactionDelegate:self];
       }
    
}

#pragma mark - Transaction Delegates

- (void)onStatusUpdate:(TransactionStatus)transactionStatus
{
    if (self.transactionDelegate!= nil && [self.transactionDelegate respondsToSelector:@selector(onStatusUpdate:)])
               {
                   [self.transactionDelegate onStatusUpdate:[C2XEnums transactionStatusFromVitalSDK:transactionStatus]];
               }
    
}
- (void)onCardholderInteractionRequested:(CardholderInteractionRequest *)request
{
    if (self.transactionDelegate!= nil && [self.transactionDelegate respondsToSelector:@selector(onCardholderInteractionRequested:)])
    {
        self.interactionResult = [CardholderInteractionResult new];        
        [self.transactionDelegate onCardholderInteractionRequested:(HpsCardholderInteractionRequest *) request];
    }
    
    [self.transactionManager sendCardholderInteractionResult:self.interactionResult];
    
}
- (void)onTransactionComplete:(TransactionResponse *)response
{
    if (self.transactionDelegate!= nil && [self.transactionDelegate respondsToSelector:@selector(onTransactionComplete:)])
    {
        [self.transactionDelegate onTransactionComplete:[HpsTerminalResponse terminalResponseFromVitalSDK:response]];
    }
}

- (void)onError:(NSError *)emvError
{
    if (self.transactionDelegate!= nil && [self.transactionDelegate respondsToSelector:@selector(onError:)])
       {
           [self.transactionDelegate onError:emvError];
       }
    
}

#pragma mark - TerminalConnectionDelegate & TerminalInfo Delegates

- (void)onConnected:(TerminalInfo *)terminalInfo
{
    if (self.deviceDelegate!= nil && [self.deviceDelegate respondsToSelector:@selector(onConnected:)])
                  {
                      [self.deviceDelegate onConnected:(HpsTerminalInfo *)terminalInfo];
                  }
}
- (void)onDisconnected
{
    
}
- (void)onTerminalInfoReceived:(nullable TerminalInfo *)terminalInfo
{
    if (self.deviceDelegate!= nil && [self.deviceDelegate respondsToSelector:@selector(onTerminalInfoReceived:)])
                     {
                         [self.deviceDelegate onTerminalInfoReceived:(HpsTerminalInfo *)terminalInfo];
                     }
}

- (void)onError
{
    if (self.deviceDelegate!= nil && [self.deviceDelegate respondsToSelector:@selector(onError:)])
    {
        [self.deviceDelegate onError:@"Device not connected"];
      
    }
}
@end

File: Pod/Classes/Terminals/C2X/HpsC2xDevice.swift
//
//  HpsC2xDevice.swift
//  Heartland-iOS-SDK
//
//  Created by Shane Logsdon on 2/7/21.
//

import Foundation

@objcMembers
public class HpsC2xDevice: NSObject, GMSClientAppDelegate, IC2xDeviceInterface {
    public var config: HpsConnectionConfig?
    public var gmsWrapper: GMSWrapper?
    public var deviceDelegate: HpsC2xDeviceDelegate?
    public var transactionDelegate: HpsC2xTransactionDelegate?
    public var peripherals: NSMutableArray?
    
    public required override init() {
        super.init()
        self.peripherals = NSMutableArray()
    }
    
    public static func initWithConfig(_ config: HpsConnectionConfig) -> Self {
        let result = self.init()
        result.config = config
        return result
    }
    
    public func initialize() {
        print("HpsC2xDevice: initialize")
        self.gmsWrapper = GMSWrapper(GMSConfiguration.fromHpsConnectionConfig(self.config!), delegate:self)
        self.scan()
    }
    
    public func scan() {
        print("HpsC2xDevice: scan")
        if let wrapper = self.gmsWrapper {
            wrapper.searchDevices()
        }
    }

    public func stopScan() {
        print("HpsC2xDevice: stopScan")
        if let wrapper = self.gmsWrapper {
            wrapper.cancelSearch()
        }
    }
    public func getDeviceInfo() {}
    public func connectDevice(_ device: HpsTerminalInfo) {
        print("HpsC2xDevice: connectDevice")
        if let wrapper = self.gmsWrapper {
            wrapper.connectDevice(device)
        }
    }
    public func processTransactionWithRequest(_ builder: HpsC2xBaseBuilder, withTransactionType transactionType: HpsTransactionType) {
        print("HpsC2xDevice: processTransactionWithRequest")
        if let wrapper = self.gmsWrapper {
            wrapper.startTransaction(builder, withTransactionType: transactionType)
        }
    }
    public func confirmAmount(_ amount: Decimal) {
        print("HpsC2xDevice: confirmAmount")
        if let wrapper = self.gmsWrapper {
            wrapper.confirmAmount(amount: amount)
        }
    }

    // mark: GMSClientAppDelegate

    public func deviceConnected() {
        print("HpsC2xDevice: deviceConnected")
        self.deviceDelegate?.onConnected() //:(HpsTerminalInfo *)terminalInfo];
    }

    public func deviceDisconnected() {
        print("HpsC2xDevice: deviceDisconnected")
        self.deviceDelegate?.onDisconnected()
    }

    public func searchComplete() {
        print("HpsC2xDevice: searchComplete")
        self.deviceDelegate?.onBluetoothDeviceList(self.peripherals ?? [])
    }

    public func deviceFound(_ device: NSObject) {
        print("HpsC2xDevice: deviceFound")
        if let peripherals = self.peripherals {
            peripherals.add(device)
        }
    }

    public func onStatus(_ status: HpsTransactionStatus) {
        print("HpsC2xDevice: onStatus")
        self.transactionDelegate?.onStatusUpdate(status)
    }


    public func onTransactionCancelled() {
        print("HpsC2xDevice: onTransactionCancelled")
    }


    public func onTransactionComplete(_ result: String, response: HpsTerminalResponse) {
        print("HpsC2xDevice: onTransactionComplete")
        print(result)
        self.transactionDelegate?.onTransactionComplete(response)
    }


    public func requestAIDSelection(_ applications: Array<AID>) {
        print("HpsC2xDevice: requestAIDSelection")
    }


    public func requestAmountConfirmation(_ amount: Decimal) {
        print("HpsC2xDevice: requestAmountConfirmation")
        self.transactionDelegate?.onConfirmAmount(amount)
    }


    public func requestPostalCode(_ maskedPan: String, expiryDate: String, cardholderName: String) {
        print("HpsC2xDevice: requestPostalCode")
    }


    public func requestSaFApproval() {
        print("HpsC2xDevice: requestSaFApproval")
    }

    public func onError(_ error: NSError) {
        print("HpsC2xDevice: onError")
        self.transactionDelegate?.onError(error)
    }
}
